modules/language-guide/pages/sharing.adoc

```diff
@@ -1,12 +1,228 @@
+
+The code below implements each of these:
+
+[source,motoko]
+----
+import Publisher "canister:pub";
+
+actor Subscriber {
+    var count: Nat = 0;
+    public func init() {
+        Publisher.subscribe(Subscriber);
+    };
+    public func notify() {
+        count += 1;
+    };
+    public func get() : async Nat {
+        count
+    };
+}
+----
+
+The actor assumes, but does not enforce, that its `init` function is
+only ever called once.  In it, the `Subscriber` actor passes a
+reference to itself, of type `actor { notify : () -> () };` (locally
+called `Subscriber` above).
+
+If called more than once, the actor will subscribe itself multiple
+times, and will receive multiple (duplicate) notifications from the
+publisher.  This fragility is the consequence of the basic
+publiisher-subscriber design we show above.  With more care, a more
+advanced publisher actor could check for duplicate subscriber actors
+and ignore them, for instance.
+
+== Sharing functions among actors
+
+In {proglang}, a `shared` actor function can be sent in a message to
+another actor, and then later called by that actor, or another one.
+
+The code shown above has been simplified for illustrative purposes.
+The full version offers additional features to the
+publisher-subscriber relationship, and uses shared functions to make
+this relationship more flexible.
+
+For instance, the notification function is _always_ designated as
+`notify`.  A more flexible design would only fix the type of `notify`,
+and permit the subscriber to choose any of its `shared` functions,
+specified in a `subscribe` message in place of (just) the actor that
+is subscribing.
+
+See the link:https://github.com/dfinity/examples/tree/master/motoko/pubsub[the full example]
+for details.
+
+In particular, suppose that the subscriber wants to
+avoid being locked into a certain naming scheme for its interface.
+What really matters is that the publisher can call _some_ function
+that the subscriber chooses.
+
+=== The keyword `shared`
+
+To permit this flexibility, an actor needs to share a single
+_function_ that permits remote invocation from another actor, not
+merely a reference to itself.
+
+The ability to share a function requires that it be pre-designated as
+`shared`, and the type system enforces that these functions follow
+certain rules around the types of data that these functions accept,
+return and over which their closures close.
+
+We omit this keyword for actor methods since implicitly, _all public
+functions of every actor are `shared`_, whether marked explicitly
+or not.
+
+More generally, a `shared` function is one that is _either_ part of
+the public interface of an actor, _or_ it is not mentioned in the
+public interface, but it does not close over the actor's mutable
+state, and it adheres to the same argument and return-type typing
+restrictions as a public actor function (e.g., no mutable data in
+the arguments or results).
```

Comment:
```suggestion
More generally, a `shared` function is one that is _either_ part of
the public interface of an actor, _or_ a private function that is explicitly declared `shared`. In either case, it adheres to the same argument and return-type typing
restrictions as a public actor function (e.g., no mutable data in
the arguments or results).
```
@crusso, do we also support local shared functions on the platform? In the interpreter, the following is fine, but I suppose we cannot do that yet:
```
> actor B { public func f() : async(shared ()->()) { shared func () {}}};
let B : actor {f : shared () -> async shared () -> ()} = {f = func}
> async (await B.f())();
async () : async<$top-level> ()
```